File: test-files/dart/vuln_bad_code.dart
// dart_security_bad_practices.dart
// Intentionally insecure + bad-practice examples for testing static analysis and reviews.
// DO NOT USE IN PRODUCTION.

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

/// 1) Hard-coded secret (embedded credential)
const String API_KEY = 'APIKEY-DO-NOT-USE-1234567890'; // secret in source

/// 2) Weak randomness for tokens (not cryptographically secure)
String generateWeakToken() {
  final r = Random(DateTime.now().millisecondsSinceEpoch); // predictable seed
  return List<int>.generate(16, (_) => r.nextInt(256)).join('-');
}

/// 3) Logging sensitive data (bad practice)
void logCredentials(String user, String pass) {
  // prints secrets to console / logs
  print('Login attempt user=$user pass=$pass');
}

/// 4) Unsafe HTTP client: disabling certificate verification
Future<String> fetchInsecure(String url) async {
  final client = HttpClient();
  // Insecure: accepts ANY certificate
  client.badCertificateCallback =
      (X509Certificate cert, String host, int port) => true;
  final req = await client.getUrl(Uri.parse(url));
  final res = await req.close();
  return await res.transform(utf8.decoder).join();
}

/// 5) Command construction for Process.run (shell injection risk)
Future<void> callShell(String filename) async {
  // Unsafe concatenation passed to shell on some platforms
  // On POSIX it's safer to pass arguments as list; this is showing poor practice.
  await Process.run('sh', ['-c', 'ls -l $filename']);
}

/// 6) SQL-like injection via string interpolation (for demo only)
String buildSql(String username) {
  // BAD: concatenating user input directly into SQL string
  return "SELECT * FROM users WHERE username = '$username';";
}

/// 7) Unsafe JSON deserialization into dynamic and using as cast
Map parseUserJson(String jsonStr) {
  // Using dynamic map without validation
  final dynamic decoded = json.decode(jsonStr);
  // Dangerous cast, may throw or allow malicious structure
  return decoded as Map;
}

/// 8) Silently catching and swallowing exceptions (lost errors)
void swallowErrors() {
  try {
    throw FormatException('boom');
  } catch (e) {
    // Bad: no logging, no rethrow, hides failure
  }
}

/// 9) Unawaited futures (fire-and-forget) causing surprising lifecycle behavior
void fireAndForget() {
  Future.delayed(Duration(seconds: 1), () => print('delayed'));
  // not awaited or tracked — OK in some contexts but often a bug
}

/// 10) Using `!` (force unwrap) unsafely with nullable types
int forceUnwrap(Map<String, int>? m) {
  // Will throw if m is null or key missing
  return m!['value']! + 1;
}

/// 11) Mutable global state (bad for concurrency and testing)
List<String> _globalCache = []; // should be encapsulated / immutable / synchronized

void mutateGlobal(String v) {
  _globalCache.add(v); // no synchronization, no validation
}

/// 12) Using `dynamic` and `as` casts to hide types (fragile)
int dangerousCast(dynamic x) {
  // developer assumes x is int but may be something else
  return (x as int) + 1;
}

/// 13) Poor file handling: predictable temp filename & race
File insecureTempWrite(String content) {
  final name = '/tmp/dart_demo_${DateTime.now().millisecondsSinceEpoch}.tmp';
  final f = File(name);
  // No atomic creation; TOCTOU possible
  f.writeAsStringSync(content);
  return f;
}

/// 14) Weak "obfuscation" of secrets (not encryption)
String xorObfuscate(String s) {
  final bytes = utf8.encode(s);
  return base64.encode(bytes.map((b) => b ^ 0xAA).toList());
}

/// 15) Unchecked user input for regexp / ReDoS potential
bool matchPattern(String input) {
  // crazy complex regex from input could be heavy — here we misuse user input
  final pattern = RegExp(input); // building regex from user-controlled string
  return pattern.hasMatch('test');
}

/// Driver that exercises some functions — call functions deliberately during tests
Future<void> main() async {
  print('Demo: Dart insecure patterns and bad practices');

  // 1
  print('Hardcoded API key: $API_KEY');

  // 2
  print('Weak token: ${generateWeakToken()}');

  // 3
  logCredentials('admin', 's3cr3t');

  // 4 - do not call with untrusted urls in CI unless sandboxed
  print(await fetchInsecure('https://plode.org'));

  // 5 - unsafe call example (commented out by default)
  await callShell('; echo hacked >/tmp/hacked.txt');

  // 6
  print(buildSql("alice'; DROP TABLE users; --"));

  // 7
  try {
    final m = parseUserJson('{"name": "bob"}');
    print('Parsed name: ${m['name']}');
  } catch (e) {
    print('parse error: $e');
  }

  // 8
  swallowErrors();

  // 9
  fireAndForget();

  // 10: force unwrap demo (catch exception so program continues)
  try {
    print('force unwrap: ${forceUnwrap(null)}');
  } catch (e) {
    print('force unwrap failed: $e');
  }

  // 11
  mutateGlobal('one');
  print('global cache size: ${_globalCache.length}');

  // 12
  try {
    print('dangerous cast: ${dangerousCast("not-an-int")}');
  } catch (e) {
    print('cast failed: $e');
  }

  // 13
  final tmp = insecureTempWrite('sensitive-data');
  print('wrote temp: ${tmp.path}');

  // 14
  print('xor obf: ${xorObfuscate("secret")}');

  // 15
  try {
    print('regex match: ${matchPattern('(a+)+\$')}');
  } catch (e) {
    print('regex error: $e');
  }
}

File: test-files/javascript/vuln_bad_code.js
// js_security_bad_practices.js
// Purpose: compact collection of common security vulnerabilities + bad coding practices
// DO NOT USE IN PRODUCTION. Use to test static analysis / CI pipelines.

const fs = require('fs');
const child = require('child_process');
const http = require('http');
const crypto = require('crypto');

// 1) Hard-coded secrets
const API_KEY = 'AKIA-VERY-SECRET-EXAMPLE';


// 2) Using Math.random() for tokens (weak RNG)
function weakToken() {
  return Math.floor(Math.random() * 1e9).toString(16);
}

// 3) eval / new Function usage (RCE risk)
function dangerousEval(userExpr) {
  // userExpr might be attacker-controlled
  return eval(userExpr); // eslint-disable-line no-eval
}

// 4) DOM sink (if used in browser): innerHTML with unsanitized input (XSS)
function unsafeDomInsert(node, html) {
  // In server-side render or electron this is dangerous too
  node.innerHTML = html; // potential XSS if html contains attacker markup
}

// 5) Child process with shell interpolation (command injection)
function listFiles(filename) {
  // Dangerous: passing a user string into shell
  child.exec(`ls -l ${filename}`, (err, stdout, stderr) => {
    if (err) console.error('exec error', err);
    console.log(stdout);
  });
}

// 6) SQL concatenation — SQLi risk (demo only; requires mysql lib)
function buildQuery(username) {
  // BAD: string concatenation instead of parameterized query
  return `SELECT * FROM users WHERE username='${username}';`;
}

// 7) Insecure HTTP server (no TLS) and CORS wildcard
const server = http.createServer((req, res) => {
  // Echoing user input to response (reflected XSS in some contexts)
  if (req.url.startsWith('/echo')) {
    const q = new URL(req.url, 'http://localhost').searchParams.get('q') || '';
    res.setHeader('Access-Control-Allow-Origin', '*'); // overly permissive CORS
    res.end(`You said: ${q}`);
  } else {
    res.end('ok');
  }
});

// 8) Storing secrets in localStorage / global (bad for web)
// localStorage.setItem('token', API_KEY); // example for browser environment

// 9) Logging sensitive data
function login(user, pass) {
  console.log(`Login attempt for ${user} with password ${pass}`); // leaks creds to logs
  return user === 'admin' && pass === 'password';
}

// 10) Path traversal (unsanitized filesystem access)
function readUserFile(userPath) {
  // BAD: reading arbitrary path supplied by user
  const base = '/var/www/data/';
  const full = base + userPath; // no normalization or validation
  return fs.readFileSync(full, 'utf8');
}

// 11) Predictable temp file name (TOCTOU)
function createTempSync() {
  const path = `/tmp/app_${process.pid}.tmp`;
  fs.writeFileSync(path, 'temporary'); // predictable, race-able
  return path;
}

// 12) Blocking synchronous I/O in async server (bad practice)
function heavySyncWork() {
  // Blocking the event loop — bad for servers
  const data = fs.readFileSync('/etc/hosts', 'utf8');
  return data;
}

// 13) Swallowing errors / empty catch
function swallowErrors() {
  try {
    JSON.parse('{"bad": }'); // will throw
  } catch (e) {
    // silently ignore -> hides bugs and potential attacks
  }
}

// 14) Using Buffer without specifying encoding & legacy APIs
function legacyBufferUse(str) {
  const b = new Buffer(str); // deprecated, insecure pattern
  return b.toString();
}

// 15) Unsafe regex constructed from user input (ReDoS)
function userRegex(input) {
  // Constructing RegExp from user input may lead to catastrophic backtracking
  const r = new RegExp(input);
  return r.test('test string');
}

// 16) Global mutable state + var usage
var GLOBAL_STORE = {}; // use of var + global mutable is bad style

// 17) Mixing callback and promise style incorrectly (promise anti-pattern)
function badPromise(cb) {
  return new Promise((resolve, reject) => {
    someAsync((err, data) => { // someAsync is assumed to exist
      if (err) return reject(err);
      resolve(data);
      if (cb) cb(null, data); // callback after resolve — confusing lifecycle
    });
  });
}

// 18) Insecure custom "encryption" (XOR obfuscation)
function obfuscateSecret(s) {
  const key = 0xAA;
  return s.split('').map(c => String.fromCharCode(c.charCodeAt(0) ^ key)).join('');
}

// Demo usage (comment dangerous calls when running in shared CI)
function main() {
  console.log('Weak token:', weakToken());

  // Dangerous: do NOT enable with untrusted input in CI unless sandboxed
  // console.log('Eval result:', dangerousEval('2+2'));

  // server.listen(8080); // start if testing CORS/XSS detection
  // listFiles('; rm -rf /tmp/test'); // Do NOT run on real systems

  try {
    console.log('Legacy buffer:', legacyBufferUse('hello'));
  } catch (e) {}

  try {
    swallowErrors();
  } catch (e) {}

  // Simulate risky logging
  login('admin', 'hunter2');

  // Create predictable temp file (for pipeline to detect)
  const tmp = createTempSync();
  console.log('Wrote tmp file:', tmp);
}

main();
